tripal/tripal
# Tripal 4 Fields: Chado Linker Fields / Chado Contact Linker Field

# New Feature

## Issue #1381

### Tripal Version: 4 alpha

## Updated Description
Convert the `ChadoContactDefault` field to dual-mode action! It will now handle contacts referenced in the base table, as it currently does, as well as being able to handle contacts referenced by a linker table. This will allow similar functionality as is provided by the Tripal 3 [Chado Linker Fields](https://github.com/tripal/chado_linker_fields) module, but without needing special linking fields.

This field should be one of the simpler "linker" fields, since all of the linker tables are simple, they do not contain additional columns. Some of the other content type linker tables also contain `type_id`, `rank`, both, or other columns.

Note that the CV term for this field in Tripal 3 was `local:contact` - here it is changed to a real term [NCIT:C47954](https://www.ebi.ac.uk/ols4/ontologies/ncit/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FNCIT_C47954) 

The following chado base tables present in Tripal 4 have a foreign key to contact:
 * study
 * arraydesign - here the foreign key column is `manufacturer_id`
 * stockcollection - this is not a default contact type

The following chado linker tables for contact exist for these content types present defined by default in Tripal 4:
 * feature_contact (gene, mRNA, QTL, SequenceVariant, GeneticMarker, Phenotypic Marker)
 * featuremap_contact (genetic_map, physical_map)
 * library_contact (dna_library)
 * project_contact (project)

## Updated Testing procedure
1. Build a new docker on this branch, since we need a new CV term for this field.
2. Create several contacts for testing. For at least one, leave the contact type blank, to test the type display, and another leave the description blank.
3. Also create an organism to test feature content types later.

### Base Table Testing
This is described in PR #1599 

### Linker Testing
1. Create a project, or any of the other linker content types listed above, and specify one contact. (dna library is missing some fields, issue #1509 so can't currently be tested) Confirm that it displays. If it has a type, it won't display yet.
2. Modify the field settings for project to display type and description. Go to `/admin/structure/bio_data/manage/project/fields` and for contact select Edit. For the token string enter this 
![Full_token_string](https://github.com/tripal/tripal/assets/8419404/29919c9e-c723-4d92-b589-44f55f90ea2b)

3. Edit the project and change the contact to a different one. Bonus points: do `select * from chado.project_contact;` to confirm the chado information is changed
4. Edit the project and add a second contact.
5. Edit the project and change the order of contacts. (the order is stored by drupal but is not saved in chado)
6. Edit the project and delete one of the contacts (change it to `-- Select --`)
7. If you add the `null` contact, it should display as `Unknown` 
![project_1](https://github.com/tripal/tripal/assets/8419404/7c926aa1-02bc-442e-ae07-8ba3453832c9)
8. repeat for a `feature` content type _e.g._ Gene, and a featuremap content type _e.g._ Physical Map

### Test the form for adding the field
1. Remove the field from one of the content types, and then manually add back the `Chado Contact` field, to test the form and ajax for base table selection, and cardinality. For example on the Project content type.
2. Go to `/admin/structure/bio_data/manage/project/fields` and delete the Contact field.
3. Now click on + Create a new field
4. Add back a "Chado Contact" field
![20231110_1](https://github.com/tripal/tripal/assets/8419404/56d05183-3968-4fa5-af56-e2b3e8aef12e)
5. Select the base table. This is where the ajax is triggered
![20231110_2](https://github.com/tripal/tripal/assets/8419404/3a1965dc-25f3-4f8f-82ab-8fd2d4cde53d)
6. The linking method should be populated with the only possibility in this case. Set cardinality to unlimited.
![20231110_3](https://github.com/tripal/tripal/assets/8419404/e16c2c22-bb51-4f2d-8f1f-9c35904b3bf9)
7. Finish defining the field and save with a token string suitable for testing
![re-add-2](https://github.com/tripal/tripal/assets/8419404/59172904-08ca-445b-8ac2-e17488f7286d)
the comment https://github.com/tripal/tripal/pull/1431#issuecomment-1809188933 talks about the token string portion now added to this form
9. Create a new project with a contact, just to confirm this works.
![re-add-3](https://github.com/tripal/tripal/assets/8419404/e3d47dff-db06-4ea6-bbdd-9ef47e1afeb3)

10. Test similarly on a content type with the contact_id in the base table, such as Study. Here the link will appear as 
![20231104_5](https://github.com/tripal/tripal/assets/8419404/a1b1528b-48be-4eea-8e55-32a0c45f0bf3)
